lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp

```diff
@@ -151,14 +186,74 @@ class LatticeValue {
   bool operator!=(const LatticeValue &other) const {
     return valueAndTag != other.valueAndTag;
   }
+  friend raw_ostream &operator<<(raw_ostream &os, const LatticeValue &dt);
 
 private:
   /// The attribute value if this is a constant and the tag for the element
   /// kind.  The attribute is always an IntegerAttr.
   llvm::PointerIntPair<Attribute, 2, Kind> valueAndTag;
 };
+
+raw_ostream &operator<<(raw_ostream &os, const LatticeValue &lattice) {
+  os << "<unknown: " << lattice.isUnknown()
+     << ", invalid: " << lattice.isInvalidValue()
+     << ", constant: " << lattice.isConstant()
+     << ", over: " << lattice.isOverdefined() << ">";
+  return os;
+}
+
+// This class defines the data structure associated with lattice values.
+// This means basically a pair of value and integer. The integer
+// represents the relative offset to the value on leaf level of type.
+class ValueAndLeafIndex {
+public:
+  ValueAndLeafIndex(Value value, unsigned leafId, bool isKnownRoot = false)
+      : valueAndFlag(value, isKnownRoot), leafId(leafId) {}
+  ValueAndLeafIndex() = default;
+  Value getValue() const { return valueAndFlag.getPointer(); }
+  unsigned getLeafIndex() const { return leafId; }
+
+  // isKnownRoot is true if we already know that the value is a root value.
+  // If this flag is true, we can skip the translation.
+  bool isKnownRoot() const { return valueAndFlag.getInt(); }
```

Comment:
What does the word "known" convey?  Can this be "isRoot"?

Also, is the terminology correct here, is this really a leaf or a root?    There can only be one root but many leaves.

Reply:
> What does the word "known" convey? Can this be "isRoot"?
Yeah, it's confusing so I'll change to "isRoot". I used isKnownRoot to avoid recomputation of the root value. 
I thought it's necessary to lookup a map to check whether the value is root or not.
But on the second thought, I think it is sufficient to check whether `isa<SubfieldOp, SubindexOp>` is true.